Makefile

```diff
@@ -12,130 +12,222 @@
+	go build -mod=readonly -ldflags="-s -w" -o cni-metrics-helper ./cmd/cni-metrics-helper
 
-# Build metrics Docker image
+# Build metrics helper agent Docker image.
 docker-metrics:
-	@docker build --build-arg arch="$(ARCH)" -f scripts/dockerfiles/Dockerfile.metrics -t "amazon/cni-metrics-helper:$(VERSION)" .
+	docker build $(DOCKER_BUILD_FLAGS) \
+		-f scripts/dockerfiles/Dockerfile.metrics \
+		-t "$(METRICS_IMAGE_NAME)" \
+		.
 	@echo "Built Docker image \"amazon/cni-metrics-helper:$(VERSION)\""
 
+# Run metrics helper unit test suite (must be run natively).
+metrics-unit-test: CGO_ENABLED=1
+metrics-unit-test: GOARCH=
 metrics-unit-test:
-	GOOS=linux CGO_ENABLED=1 go test -v -cover -race -timeout 10s ./cmd/cni-metrics-helper/metrics/...
+	go test -v -cover -race -timeout 10s -mod=readonly \
+		./cmd/cni-metrics-helper/metrics/...
 
+# Run metrics helper unit test suite in a container.
 docker-metrics-test:
-	docker run -v $(shell pwd):/usr/src/app/src/github.com/aws/amazon-vpc-cni-k8s \
-		--workdir=/usr/src/app/src/github.com/aws/amazon-vpc-cni-k8s \
-		--env GOPATH=/usr/src/app \
-		golang:1.10 make metrics-unit-test
-
-# Build both CNI and metrics helper
-all: docker docker-metrics
+	docker run $(DOCKER_RUN_FLAGS) \
+		-v $(shell pwd -P):/src --workdir=/src \
+		-e GOARCH -e GOOS -e GO111MODULES \
+		$(GOLANG_IMAGE) \
+		make metrics-unit-test
 
+# Generate descriptors for supported ENI configurations.
 generate-limits:
-	go run pkg/awsutils/gen_vpc_ip_limits.go
-
-# golint
-# To install: go get -u golang.org/x/lint/golint
+	go run -mod=readonly pkg/awsutils/gen_vpc_ip_limits.go
+
+# Fetch portmap the port-forwarding management CNI plugin
+portmap: FETCH_VERSION=0.7.5
+portmap: FETCH_URL=https://github.com/containernetworking/plugins/releases/download/v$(FETCH_VERSION)/cni-plugins-$(GOARCH)-v$(FETCH_VERSION).tgz
+portmap: VISIT_URL=https://github.com/containernetworking/plugins/tree/v$(FETCH_VERSION)/plugins/meta/portmap
+portmap:
+	@echo "Fetching portmap CNI plugin v$(FETCH_VERSION) from upstream release"
+	@echo
+	@echo "Visit upstream project for portmap plugin details:"
+	@echo "$(VISIT_URL)"
+	@echo
+	curl -L $(FETCH_URL) | tar -z -x ./portmap
+
+# Run all source code checks.
+check: check-format lint vet
+
+# Run golint on source code.
+#
+# To install:
+#
+#   go get -u golang.org/x/lint/golint
+#
+lint: LINT_FLAGS = -set_exit_status
 lint:
-	golint pkg/awsutils/*.go
-	golint cmd/routed-eni-cni-plugin/*.go
-	golint cmd/routed-eni-cni-plugin/driver/*.go
-	golint pkg/k8sapi/*.go
-	golint pkg/networkutils/*.go
-	golint pkg/ipamd/*.go
-	golint pkg/ipamd/*/*.go
-
-# go vet
+	@command -v golint >/dev/null || { echo "ERROR: golint not installed"; exit 1; }
+	find . -depth -mindepth 1 \
+	  -type d \
+	  -not -path ./rpc -not -path ./pkg -not -path ./cmd -not -path './.*' \
+	  -not -name mocks -not -name mock \
+	  -not -name 'mock_*' -not -name 'mocks_*' \
```

Comment:
the above includes the `docs/`, `testdata/`, `misc/`, `/config` and `/scripts` directories which should be excluded from Go linting.

Probably better to just search for Go files and exclude mocks:

```
find . -name *.go -not -name *mock*
```

but be aware that there are linter failures in the tree's current code, so we'll need to clean these up before adding lint to the `check` target that gates CI:

```
jaypipes@udb42383a8e5155:~/go/src/github.com/aws/amazon-vpc-cni-k8s$ for f in `find . -name *.go -not -name *mock*`; do GO111MODULE=on golint $f; done
./pkg/awsutils/awsutils.go:194:1: exported method ENIMetadata.PrimaryIPv4Address should have comment or be unexported
./pkg/ec2metadatawrapper/ec2metadatawrapper.go:17:6: type HttpClient should be HTTPClient
./pkg/ipwrapper/ip.go:23:6: exported type IP should have comment or be unexported
./pkg/ipwrapper/ip.go:30:1: exported function NewIP should have comment or be unexported
./pkg/eniconfig/eniconfig.go:50:6: exported type ENIConfig should have comment or be unexported
./pkg/eniconfig/eniconfig.go:55:5: exported var ErrNoENIConfig should have comment or be unexported
./pkg/eniconfig/eniconfig.go:58:6: type name will be used as eniconfig.ENIConfigController by other packages, and that stutters; consider calling this Controller
./pkg/eniconfig/eniconfig.go:68:6: type name will be used as eniconfig.ENIConfigInfo by other packages, and that stutters; consider calling this Info
./pkg/eniconfig/eniconfig.go:162:1: exported method ENIConfigController.Getter should have comment or be unexported
./pkg/utils/ttime/ttime.go:14:6: exported type Timer should have comment or be unexported
./pkg/utils/retry/retry.go:27:6: func name will be used as retry.RetryWithBackoff by other packages, and that stutters; consider calling this WithBackoff
./pkg/utils/retry/retry.go:35:6: func name will be used as retry.RetryWithBackoffCtx by other packages, and that stutters; consider calling this WithBackoffCtx
./pkg/utils/retry/retry.go:57:6: func name will be used as retry.RetryNWithBackoff by other packages, and that stutters; consider calling this NWithBackoff
./pkg/utils/retry/retry.go:66:6: func name will be used as retry.RetryNWithBackoffCtx by other packages, and that stutters; consider calling this NWithBackoffCtx
./pkg/utils/retry/backoff.go:23:6: exported type Backoff should have comment or be unexported
./pkg/utils/retry/backoff.go:28:6: exported type SimpleBackoff should have comment or be unexported
./pkg/utils/retry/backoff.go:51:1: exported method SimpleBackoff.Duration should have comment or be unexported
./pkg/utils/retry/backoff.go:59:1: exported method SimpleBackoff.Reset should have comment or be unexported
./pkg/utils/retry/errors.go:21:6: exported type Retriable should have comment or be unexported
./pkg/utils/retry/errors.go:25:6: exported type DefaultRetriable should have comment or be unexported
./pkg/utils/retry/errors.go:29:1: exported method DefaultRetriable.Retry should have comment or be unexported
./pkg/utils/retry/errors.go:33:1: exported function NewRetriable should have comment or be unexported
./pkg/utils/retry/errors.go:39:6: exported type RetriableError should have comment or be unexported
./pkg/utils/retry/errors.go:44:6: exported type DefaultRetriableError should have comment or be unexported
./pkg/utils/retry/errors.go:49:1: exported function NewRetriableError should have comment or be unexported
./pkg/utils/retry/errors.go:56:6: exported type AttributeError should have comment or be unexported
./pkg/typeswrapper/client.go:20:6: exported type CNITYPES should have comment or be unexported
./pkg/typeswrapper/client.go:27:1: exported function New should have comment or be unexported
./pkg/apis/crd/v1alpha1/register.go:17:2: exported var SchemeBuilder should have comment or be unexported
./pkg/apis/crd/v1alpha1/types.go:9:6: exported type ENIConfigList should have comment or be unexported
./pkg/apis/crd/v1alpha1/types.go:17:6: exported type ENIConfig should have comment or be unexported
./pkg/apis/crd/v1alpha1/types.go:24:6: exported type ENIConfigSpec should have comment or be unexported
./pkg/apis/crd/v1alpha1/types.go:28:6: exported type ENIConfigStatus should have comment or be unexported
./pkg/rpcwrapper/client.go:21:6: exported type RPC should have comment or be unexported
./pkg/rpcwrapper/client.go:27:1: exported function New should have comment or be unexported
./pkg/cri/cri.go:27:6: exported type APIs should have comment or be unexported
./pkg/cri/cri.go:31:6: exported type Client should have comment or be unexported
./pkg/cri/cri.go:33:1: exported function New should have comment or be unexported
./pkg/cri/cri.go:37:1: exported method Client.GetRunningPodSandboxes should have comment or be unexported
./pkg/ipamd/ipamd.go:188:1: comment on exported type ReconcileCooldownCache should be of the form "ReconcileCooldownCache ..." (with optional leading article)
./pkg/ipamd/ipamd_test.go:275:2: var warmIpTarget should be warmIPTarget
./pkg/grpcwrapper/client.go:20:6: exported type GRPC should have comment or be unexported
./pkg/grpcwrapper/client.go:26:1: exported function New should have comment or be unexported
./pkg/k8sapi/discovery.go:120:1: comment on exported method Controller.DiscoverCNIK8SPods should be of the form "DiscoverCNIK8SPods ..."
./pkg/ec2wrapper/ec2wrapper.go:1:1: package comment should be of the form "Package ec2wrapper ..."
./pkg/ec2wrapper/ec2wrapper.go:30:1: comment on exported function NewMetricsClient should be of the form "NewMetricsClient ..."
./pkg/ec2wrapper/client.go:23:6: exported type EC2 should have comment or be unexported
./pkg/ec2wrapper/client.go:37:1: exported function New should have comment or be unexported
./pkg/nswrapper/ns.go:20:6: exported type NS should have comment or be unexported
./pkg/nswrapper/ns.go:27:1: exported function NewNS should have comment or be unexported
./cmd/routed-eni-cni-plugin/cni.go:82:2: don't use ALL_CAPS in Go names; use CamelCase
./cmd/routed-eni-cni-plugin/cni.go:85:2: don't use ALL_CAPS in Go names; use CamelCase
./cmd/routed-eni-cni-plugin/cni.go:88:2: don't use ALL_CAPS in Go names; use CamelCase
./cmd/routed-eni-cni-plugin/cni.go:288:10: if block ends with a return statement, so drop this else and outdent its block
./cmd/routed-eni-cni-plugin/cni.go:301:2: var deletedPodIp should be deletedPodIP
```